Assets/Scripts/Hit.cs

```diff
@@ -0,0 +1,177 @@
+using UnityEngine;
+using UnityEngine.Events;
+using System.Collections.Generic;
+using System.Linq;
+
+// Interface for objects that can hit others
+public interface IHitter
+{
+    void ApplyDamage(GameObject target);
+}
+
+public class Hit : MonoBehaviour, IHitter
+{
+    [Header("Damage Settings")]
+    [SerializeField] private float damage = 10f;
+
+    [Header("Hit Behavior")]
+    [SerializeField] private bool destroyOnHit = false;
+    [SerializeField] private int hitMaxTimes = 1;
+    [SerializeField] private float hitCooldown = 0.5f;
+    [SerializeField] private bool debugMode = false;
+
+    [Header("Events")]
+    public UnityEvent<GameObject> OnHitObject;
+    public UnityEvent<GameObject> OnBeforeDestroy;
+
+    private Dictionary<GameObject, int> hitCounts = new Dictionary<GameObject, int>(); // Track hit counts
+    private Dictionary<GameObject, float> hitCooldowns = new Dictionary<GameObject, float>();
```

Comment:
these are unbounded and could potentially grow, is it worth having a cleanup or pooling these?